### PR TITLE
qtile: add migrate command

### LINKED ISSUE #1773: idea: qtile-config {check,migrate}
For a long time we've had issues with configs:

1. people have syntax/logic errors and have to restart a bunch
2. we want to make changes to the internals and people are using them
3. people use the wrong kind of values for config variables, but it works "well enough", except doesn't get the desired behavior
4. etc.

In particular, we've long struggled with number 2, and have no real conclusions about what to support for a 1.0 API. For example, we (rightly so) got rid of a bunch of camel case in our codebase, but it broke everyone's configs who was messing with the internals. I've argued that everything that isn't a `cmd_` is not a public API and we reserve the right to change, but it's still annoying for people to have to update their configs.

Further, r.e. numbers 1 & 3, we've often told people "just run your config via python3 config.py" or (when we get type annotations working) "run pytype on your config". But there are additional checks we could/should have people do (see e.g. #1772) that neither of those will accomplish.

So, what if we introduce a new command `qtile-config`. Initially I'm imagining two subcommands, `check` and `migrate`. `check` would run a type checker, our checks from #1772, and anything else we want (deprecation warnings, etc.). `migrate` would do a best-effort migration of code. In general re-writing code is undecidable, but most of our changes are things like "add another argument to this function" or "rename fooBar to foo_bar", which we could do with reasonable accuracy, and if your config doesn't get auto-migrated, well, too bad.

Then, after 1.0, we allow any API changes where someone can write a migrate blurb that would be reasonably successful at migrating most configs. This means that 1. users don't have do this themselves and 2. we can keep refactoring our code for e.g. wayland.

Once we had this code, we could do other things too, like run `migrate` on qtile start so users' configs are automatically upgraded.

There is an issue of dependencies, but we can make the type checking deps for `qtile-config check` optional, so people don't have to install them if they don't want, and just print a big warning if they're not available.

I'm happy to work on an implementation of this, but I'd just like to see if anyone else has any good ideas before starting :)

### PR DESCRIPTION
**Pull request by tych0**
Here's an implementation of the proposed `qtile migrate` command.

For now, I've used pybowler. While they say in their docs that their API
may not be stable, it seems mostly unchanged since 2016 when it was
introduced. I played around with other refactoring packages like rope, and
pybowler ends up being much simpler for the most part.

I tried to write a variety of migrations from our last release that people
complained about. One I looked at briefly but couldn't figure out was how
to automatically remove the qtile arg mouse_callbacks in widget
constructors. I think it's probably reasonably easy if anyone is
interested.

One thing I didn't do was write any sort of policy about what kinds of
requirements we want to place on API breaking changes, or what we want to
state as our official API. For now, I'd like to require any API breaking
changes at least do a good faith effort at a migration, and let's see how
much of a burden it places. But others' thoughts are welcome on this policy
:)

I'm going to say this:

Closes #1773

although the conversation about API stability continues...

Signed-off-by: Tycho Andersen <tycho@tycho.pizza>

### CODE REVIEW

File: libqtile/scripts/migrate.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            bowler.Query(config)
+            .select_module(fro)
+            .rename(to)
+        )
+    MIGRATIONS.append(f)
+
+
+def do_migrate(args):
+    if "bowler" not in sys.modules:
+        print("bowler can't be found, not mirating config file")
```
Comment (ramnes) [inline libqtile/scripts/migrate.py:88]:
```suggestion
        print("bowler can't be found, not migrating config file")
```

**Review by ramnes — APPROVED** [review]
Awesome!

File: libqtile/scripts/migrate.py (hunk 1 above)
Comment (tych0) [reply libqtile/scripts/migrate.py]: Good catch, thanks!

File: libqtile/scripts/migrate.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            bowler.Query(config)
+            .select_module(fro)
+            .rename(to)
+        )
+    MIGRATIONS.append(f)
+
+
+def do_migrate(args):
+    if "bowler" not in sys.modules:
+        print("bowler can't be found, not mirgating config file")
```
Comment (m-col) [inline libqtile/scripts/migrate.py:88]: Third time's a charm?

**Comment by m-col** [comment]
Admittedly I have been a bit skeptical about how feasible this was but this looks really cool!

The specific changes that we add to command only make sense for migrating from one Qtile release to another (which I assume is ultimately the purpose of the command, less so for those git-tracking users). Therefore it might be worth labelling each migration with the relevant release (e.g. `"for 0.17.0"` in the docstring). This way we can remove them in a release or two.

> For now, I'd like to require any API breaking
changes at least do a good faith effort at a migration, and let's see how
much of a burden it places.

This seems like a sensible approach. I just had a quick read about bowler and the API looks pretty nice, but having examples within our own code will definitely help others who need to add them for their own API-breaking changes. The migrate command itself that you have implemented looks pretty straightforward and understandable.

Are the pyi files for mypy or for bowler?

File: libqtile/scripts/migrate.py (hunk 2 above)
Comment (tych0) [reply libqtile/scripts/migrate.py]: bleh!

**Comment by tych0** [comment]
> Are the pyi files for mypy or for bowler?

For our mypychecks, bowler doesn't export type info unfortunately.

**Comment by tych0** [comment]
I'm gonna go ahead and merge this for now. We should figure out if/how we want to make this automatic (only when config fails to load, every time, etc) and also it would be great if other folks play around with bowler. I'm definitely not married to it as a library to use for this (and perhaps there's some things we could build our own crude parsers for, etc). But no way to find out answers to these questions until people start playing with it.

**Pull request merged**